node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -631,6 +632,45 @@ async fn disconnect_peer(sender: &mut impl overseer::CollatorProtocolSenderTrait
 		.await
 }
 
+/// Check if `PendingCollation` can be collated.
+async fn can_collate(
+	pc: &PendingCollation,
+	sender: &mut impl overseer::CollatorProtocolSenderTrait,
+) -> bool {
+	let PendingCollation { relay_parent, para_id, peer_id, .. } = pc;
+	let mc = polkadot_node_subsystem_util::request_availability_cores(*relay_parent, sender)
+		.await
+		.await
+		.ok()
+		.and_then(|x| x.ok());
+
+	let cores = match mc {
+		Some(c) => c,
+		None => {
+			gum::debug!(
+				target: LOG_TARGET,
+				?relay_parent,
+				"Failed to query runtime API for relay-parent",
+			);
+
+			return false
+		},
+	};
+
+	for core in cores {
+		match core {
+			CoreState::Scheduled(sc) if sc.para_id == *para_id => {
+				let op_peer_id = OpaquePeerId::new(peer_id.to_bytes());
+
+				return sc.collator_restrictions.can_collate(&op_peer_id)
```

Comment:
Ideally can_collate would take a `PeerId` and do the conversion to opaque internally. I would really like to hide this implementation detail as much as possible.

Reply:
Isn't the issue that we can't have `PeerId` in the runtime in the first place and that's why we use `OpaquePeerId`? Hence, we do need to convert here? I guess we could have functions to convert between the two to make it look nicer, though.